### ISSUE TITLE
AmazonS3Client.GetObjectAsync is not threadsafe

### ISSUE THREAD

**Issue by LucVL** [aws/aws-sdk-net#1971]
## Description
When calling AmazonS3Client.GetObjectAsync simultaneously from different threads, spurious exceptions can occur that suggest threading issues.

## Reproduction Steps
* List the objects in a bucket.
* For each object listed, start a thread to get the object and handle the contents
* In my test the bucket contains 3 files of about 60GiB that are normally read through the ResponseStream of the response object returned by GetObjectAsync. Roughly half of the time an exception is thrown.

```c#
using Amazon.S3;
using Amazon.S3.Model;

namespace AmazonS3Client_GetObjectAsync;

public class Program
{
    public static Task<int> Main(string[] args)
    {
        var bucketName = args[0];
        var prefix = args[1];

        return RunAsync(bucketName, prefix);
    }

    private static async Task<int> RunAsync(string bucketName, string prefix)
    {
        using var s3Client = new AmazonS3Client();

        var listObjectsRequest = new ListObjectsRequest
        {
            BucketName = bucketName,
            Prefix = prefix,
        };

        var listObjectsResponse = await s3Client.ListObjectsAsync(listObjectsRequest);

        var tasks = new List<Task>();

        foreach (var s3Object in listObjectsResponse.S3Objects)
        {
            if (s3Object.Key == prefix)
                continue;

            var asyncFunc = () => HandleS3ObjectAsync(s3Client, s3Object);

            tasks.Add(Task.Run(asyncFunc));
        }

        await Task.WhenAll(tasks);

        return 0;
    }

    private static async Task HandleS3ObjectAsync(AmazonS3Client s3Client, S3Object s3Object)
    {
        Console.WriteLine($"{s3Object.Key}: {s3Object.Size} bytes");

        var request = new GetObjectRequest
        {
            BucketName = s3Object.BucketName,
            Key = s3Object.Key,
        };

        using var response = await s3Client.GetObjectAsync(request);

        // intentionally left blank
    }
}
```

### Sample exception thrown

* InvalidOperationException: Collection was modified; enumeration operation may not execute
* System.IndexOutOfRangeException: Index was outside the bounds of the array.
* Amazon.S3.AmazonS3Exception: The request signature we calculated does not match the signature you provided. Check your key and signing method.

See [Exceptions.txt](https://github.com/aws/aws-sdk-net/files/7958874/Exceptions.txt)

### Environment
* SDK Version: AWSSDK.S3, both 3.7.7.5 and 3.7.7.16
* OS Info: Windows Server 2016 on an AWS EC2 m5n.2xlarge instance (arch: AMD64)
* Build Environment Visual Studio 2022 (64-bit, 17.0.5)
* Targeted .NET Platform:  net6.0, Self-contained, win-x64

## Workaround

Use a common SemaphoreSlim to prevent simultaneous execution of GetObjectAsync.

```c#
        await s_SemaphoreSlim.WaitAsync();

        try
        {
            response = await s3Client.GetObjectAsync(request);
        }
        finally
        {
            s_SemaphoreSlim.Release();
        }
```

**Comment by LucVL** [comment]
I tried using a separate S3 client per request and that didn't change anything (See [Exceptions 20220131.txt](https://github.com/aws/aws-sdk-net/files/7969985/Exceptions.20220131.txt): I tried 3 runs, all failed with 2 different exceptions)

I did another test by targeting the original code to .Net 5 (after changing the code a bit to remove C#10 specific constructs) and then no exceptions where thrown after 10 runs.

So it does seem to be linked to the issue you mentioned and the change in .Net 6 ([Remove outdated lock from ParseRawHeaderValues #54130](https://github.com/dotnet/runtime/pull/54130))

**Comment by plachor** [comment]
@ashishdhingra  many days past since this was reported, change on Microsoft implantation seems reosonble and does not look like they will change any thing for .NET6.  What are your plans? Any one will grab this and fix on your side?

.NET6 is LTS release so seems like this must be addressed.

**Comment by Travonka** [comment]
Hey all! Does this issue still relevant?

**Comment by LucVL** [comment]
@Travonka: Has there been any change in the aws sdk to correct this?
The random nature of the exceptions thrown point to a threading issue. The bug is probably related to the aws sdk depending on an undocumented 'implementation feature' of the .Net 5 (There used to be a lock in ParseRawHeaderValues: [Remove outdated lock from ParseRawHeaderValues](https://github.com/dotnet/runtime/pull/54130))
I provided detailed steps to reproduce.
For now all my projects that need it use the workaround I provided.
If I find the time, I'll rerun the tests, but I can't guarantee any time soon...

**Comment by Travonka** [comment]
@LucVL , hey, thank you for answering. The reasong I'm asking is that  I'm trying to reproduce this issue using exactly your code, but i did not manage to catch the error.

**Comment by plachor** [comment]
In our case we are running under Kubernetes, we observe those errors like once every 12k successful calls but perhaps this correlates with other calls happening in background not directly related with S3 but occurring along in same process and therefore increasing odds.  Before .NET6 we never observed them.

**Comment by Travonka** [comment]
Oh.
once in 12k.
I guess that’s might be the reason why I didn’t manage to catch it.
I’l do some load testing on it

**Comment by LucVL** [comment]
@Travonka When I filed the report, test runs failed about half the time.
I reran the tests now and didn't have a single failure, not even with the original binaries I still had present on a test machine, so I don't understand how it got corrected.
One wild guess would be that the S3 server added some headers so the client doesn't need the default headers any more.

I noticed there's another pull request ([Make concurrent reads on HttpHeaders thread-safe #68115](https://github.com/dotnet/runtime/pull/68115) merged to main on the dotnet/runtime repo, but that's not public yet.
